x-pack/plugins/siem/server/lib/detection_engine/signals/filter_events_with_list.ts

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import { get } from 'lodash/fp';
+import { Logger } from 'src/core/server';
+
+import { ListItemArraySchema } from '../../../../../lists/common/schemas/response';
+import { Type as ListValueType } from '../../../../../lists/common/schemas/common';
+import { ListClient } from '../../../../../lists/server';
+import { SignalSearchResponse } from './types';
+import { RuleAlertParams } from '../types';
+
+interface FilterEventsAgainstList {
+  listClient: ListClient;
+  exceptionsList: RuleAlertParams['exceptions_list'];
+  logger: Logger;
+  eventSearchResult: SignalSearchResponse;
+}
+
+export const filterEventsAgainstList = async ({
+  listClient,
+  exceptionsList,
+  logger,
+  eventSearchResult,
+}: FilterEventsAgainstList): Promise<SignalSearchResponse> => {
+  try {
+    if (exceptionsList == null) {
+      return eventSearchResult;
+    }
+
+    // grab the signals with values found in the given exception lists.
+    const filteredHitsPromises = exceptionsList.map(async exceptionItem => {
+      // acquire the list values we are checking for.
+      const valuesOfGivenType = eventSearchResult.hits.hits
```

Comment:
I would remove and avoid any lodash get's at this phase since we have many more TypeScript tooling at our disposal in a few ways and lodash get likes to return `any` which messes things up with strong types.

Making an assumption below that you changed your `SignalSource` to use `unknown` from my comments below you would do something like this now:

```ts
const valuesOfGivenType = eventSearchResult.hits.hits
        .map(searchResultItem => {
          return searchResultItem._source[exceptionItem.field];
        })
        .filter(searchResultItem => searchResultItem != null);
      const valueSet = new Set<string>(valuesOfGivenType as string[]); // make them small
```

You do have an `as cast` but want to point something out with defining the majority of our types now using `io-ts` is that we have a built in `.is` we can probably use if I didn't mess up my code when I wrote io-ts.

For example if you need to check if `exceptionItem.values[0].name` is a type of `type` you can do this:

```ts
import { Type as ListValueType, type } from '../../../../../lists/common/schemas/common';
...
      if (type.is(exceptionItem.values[0].name) && exceptionItem.values[0].id != null) {
        const listSignals = await listClient.getListItemByValues({
          listId: exceptionItem.values[0].id,
          type: exceptionItem.values[0].name,
          value: [...valueSet],
        });
      }
```

which is the nice thing about `io-ts` types they offer a built in type guard automatically so you don't have to fuss around with writing that. I would try to remove this:

```ts
listId: exceptionItem.values[0].id!,
```
or do fast validations and throws maybe when you know the shape should not be invalid or you just log some warn/errors and return early if things don't validate or a big if/else block, etc... Up to you on that, but turning off that type assertion is problematic usually.
By

Reply:
`type.is` is awesome thanks for that! As for the lodash get - because the `exceptionItem.field` is a string representing a flattened path (like `host.name`) to the value in the source event, I need a way to access them outside of `_source[exceptionItem.field];` as that will just return undefined since the `host.name` key doesn't exist at the `_source` level.

Reply:
Fixed [here](6ebc0cdd8397706dd57c45570661610de85d7de1)